src/app/introduction/picture-puzzle.component.ts

```diff
@@ -0,0 +1,384 @@
+                                        this.webcamStream = stream;
+                                    })
+        }
+    }
+
+    snapshot() {
+        this.webcamState = this.webcamStates.CAPTURED;
+        if (this.webcamStream.stop) {
+            this.webcamStream.stop();
+        } else {
+            this.webcamStream.getTracks().forEach(function(track) { track.stop() })
+        }
+        this.webcamButtonText = 'Take another photo';
+        this._pieceWidth = Math.floor(this._width / this.PUZZLE_DIFFICULTY)
+        this._pieceHeight = Math.floor(this._height / this.PUZZLE_DIFFICULTY)
+        this._puzzleWidth = this._pieceWidth * this.PUZZLE_DIFFICULTY;
+        this._puzzleHeight = this._pieceHeight * this.PUZZLE_DIFFICULTY;
+        this.setCanvas();
+        this._stage.drawImage(this._video, 0, 0, this._width, this._height);
+        this.activityComplete = true;
+        swal(
+            'Good job!',
+            'You completed this activity!',
+            'success'
+        );
+    }
+
+    ngOnInit() {
+        this.webcamButtonText = 'Take Photo';
+        this.checkProgress();
+    }
+
+    upload() {
+        const formData: any = new FormData();
+        const files: Array<File> = this.filesToUpload;
+        if (this.webcamState === this.webcamStates.CAPTURED) {
+            const camData = {base64: this._canvas.toDataURL('image/jpeg')}
+            this._http.post('http://localhost:3000/api/uploadCam', camData)
+                .map(files => camData)
+                .subscribe(files => {
+                    swal(
+                        'Success!',
+                        'Profile picture updated successfully!',
+                        'success'
+                    );
+                 });
+        } else {
+            formData.append('uploads[]', files, 'profile_picture');
+            this._http.post('http://localhost:3000/api/upload', formData)
+                .map(files => files.json())
+                .subscribe(files => {
+                    swal(
+                        'Success!',
+                        'Profile picture updated successfully!',
+                        'success'
+                    );
+                });
+        }
+    }
+
+    fireFileButtonClick() {
+        if (this.webcamStream) {
+            if (this.webcamStream.stop) {
+                this.webcamStream.stop();
+            } else {
+                this.webcamStream.getTracks().forEach(function(track) {track.stop()})
+            }
+        }
+        this.webcamState = this.webcamStates.PAGE_LOAD;
+        this.webcamButtonText = 'Take Photo';
+        const file = document.getElementById('file').click();
+    }
+
+    handleInputChange(e) {
+        const file = e.dataTransfer ? e.dataTransfer.files[0] : e.target.files[0];
+        const pattern = /image-*/;
+        const reader = new FileReader();
+
+        if (!file.type.match(pattern)) {
+            alert('invalid format');
```

Comment:
`alert`, `confirm` are extremely ugly (blocks js execution & stops everything entirely) & is not suggested by js.
you should use modals intead